### PR TITLE
[Performance][MoE] Reuse a persistent pad buffer in the MC2 prepare path

### PR DESCRIPTION
**Pull request by MirkoDeVita98**
### What this PR does / why we need it?

`PrepareAndFinalizeWithMC2.prepare` pads `hidden_states` and `router_logits` up to `padded_num_tokens` with `nn.functional.pad`, which allocates a new tensor and zero-fills it on every call.

Since

```text
padded_num_tokens = ceil(max_tokens_across_dp / tp_size) * tp_size
```

the padded shape depends on the TP width and not on the data, so it is constant for a given captured shape and the allocation can be hoisted into a persistent buffer.

This costs two extra kernels per MoE layer per forward whenever `pad_size > 0`, i.e. whenever the token count is not an exact multiple of `tp_size` — **7 of every 8 batch sizes at `tp_size=8`**.

Profiled with `torch_npu.profiler` on Qwen3-30B-A3B, 8x Atlas 800I A2, EP=8, batch 1, ACL graph captured, rank 0:

| kernel | before | after |
|---|--:|--:|
| `aclnnConstantPadNd_PadV3AiCore_PadV3` | 945.2 us/step | 0.0 (0 calls) |
| `aclnnConstantPadNd_PadV3AiCore_MemSet` | 808.4 us/step | 0.0 (0 calls) |
| all padding kernels | 36.7 us/layer | 0.20 us/layer |

At 18.6 us per call for 32 KB of data, this is launch overhead rather than bandwidth, so the absolute cost is largely independent of how many rows are padded.

The AllGather path does not pay it at all: its pad size is `max_tokens_across_dp - num_tokens`, which is 0 when `dp_size == 1`.

Where `pad_size == 0`, the `if pad_size > 0` guard means `_pad_to` is never called, so the change is a no-op on those shapes rather than a different code path.

**Correctness of the padded rows.** `nn.functional.pad` guarantees the padded rows are zero, and that is preserved here because the cache key includes the unpadded row count as well as the padded length.

Several batch sizes map to the same `padded_num_tokens` (at `tp_size=8`, batches 1..8 all pad to 8 rows), so a buffer keyed only on the padded length would keep a larger batch's rows where a smaller one expects zeros. Keying on both means the padded rows are never written for a given buffer and stay zero from allocation onwards, at a cost of one small buffer per captured shape.

This is not hypothetical: `mc2_mask` is passed to the dispatch op only when `global_bs == 0`, so on the other path the op does read those rows.

### Does this PR introduce _any_ user-facing change?

No. Behaviour is unchanged; only the padding allocation strategy differs.

### How was this patch tested?

Qwen3-30B-A3B (48 layers, 128 experts, top-8, bf16), 8x Atlas 800I A2 (910B2), EP=8/TP=8, `cudagraph_mode: FULL_DECODE_ONLY`, MC2 comm path.

**Correctness** — paired per-token logprob over 510 prefill tokens:

 ```text
unpatched  ppl 1.7135
patched    ppl 1.7128
paired delta +0.000398 +/- 0.002108 nats (n=510, 0.19 sigma)
```

**Decode latency, batch 1** — 32 prompts served one at a time, 128 generated tokens, 3 rounds with the arms rotated within each round to cancel thermal drift:

| round | ITL unpatched | ITL patched | delta |
|--:|--:|--:|--:|
| 1 | 18.3972 ms | 17.6319 ms | -4.16% |
| 2 | 18.4899 ms | 17.5666 ms | -4.99% |
| 3 | 18.5019 ms | 17.7990 ms | -3.80% |

```text
ITL  -0.7972 ms (-4.32%), se 0.0656 -> 12.2 sigma
TTFT 176.36 -> 176.56 ms (unchanged)
```

AllGather runs at 16.9456 ms ITL in the same rounds, so MC2 goes from +8.95% to +4.25% against it.

**Decode throughput across batch sizes** — 3 reps per arm, medians:

| batch | pad_size | unpatched | patched | delta |
|--:|--:|--:|--:|--:|
| 1 | 7 | 46.2 tok/s | 48.5 tok/s | +5.0% |
| 4 | 4 | 156.9 tok/s | 165.8 tok/s | +5.7% |

The `pad_size == 0` row is the control: the padding path is not entered, and the residual +1.4% sets the noise floor for this measurement, against which the +5.0% and +5.7% stand.

Not yet measured: a large batch with `pad_size > 0` under graph capture. Batches that are not a multiple of 8 are only captured when `max_num_seqs` is raised to the next multiple, because vLLM's ladder is `[1, 2, 4] + range(8, max_capture + 1, 8)`; otherwise, the step exceeds every captured size and falls back to eager, where this padding is ~1% of the step and invisible.

The reasoning above (launch-bound, so roughly batch-independent) suggests the win persists, but it is reasoning rather than a measurement.

### CODE REVIEW

File: vllm_ascend/ops/fused_moe/prepare_finalize.py [hunk 1]
```diff
@@ -248,6 +248,32 @@ def _restore_tp_across_dp(self):
...
+        cache = getattr(self, "_pad_cache", None)
+        if cache is None:
+            cache = self._pad_cache = {}
+        key = (tuple(t.shape[1:]), t.dtype, target)
+        buf = cache.get(key)
+        if buf is None or buf.device != t.device:
+            buf = torch.zeros((target, *t.shape[1:]), dtype=t.dtype, device=t.device)
+            cache[key] = buf
+        buf[: t.shape[0]].copy_(t)
+        return buf
```
[the remark this answers is not in the document]
Comment (MirkoDeVita98) [reply vllm_ascend/ops/fused_moe/prepare_finalize.py]: That the concern is valid, was fixed in the current revision by including the unpadded row count in the cache key, and that the reason it matters isn't hypothetical: mc2_mask reaches the dispatch op only when global_bs == 0, so on the other path the op genuinely reads those rows.
